scripts/macos/RenameMacUserNameAndHomeDirectory.sh

```diff
@@ -1,61 +1,137 @@
+
+ The script updates the users record name, RealName (displayName), and home directory
+
+ If the user recieving the name change is signed in they will be signed out. 
+
+ Example usage: 
+
+ sudo sh RenameMacUserNameAndHomeDirectory.sh cat dog
+
+ Above example would rename account cat to dog
+
+ NOTE: SCRIPT MUST BE RUN AS ROOT!
+'
+
 abort() {
 	errString=${*}
 	echo "$errString"
 	exit 1
 }
 
-if [[ ${#} -ne 2 ]]
-then
+if [[ ${#} -ne 2 ]]; then
 	echo "Usage: $0 oldUserName newUserName"
 	exit 1
 fi
 
 oldUser=$1
 newUser=$2
 
-if [[ -z "${newUser}" ]]
-then
-	abort "New user name must not be empty!"
+#Verify valid username
+if [[ -z "${newUser}" ]]; then
+	echo "New user name must not be empty!"
+	exit 1
+fi
+
+#Test to ensure account update is needed
+if [ "$oldUser" = "$newUser" ]; then
+	echo "No updates needed"
+	exit 0
 fi
 
-origHomeDir=`dscl . -read /Users/${oldUser} NFSHomeDirectory | awk '{print $2}' -`
+#Query existing user accounts
+readonly existingUsers=($(dscl . -list /Users | grep -Ev "^_|com.*|root|nobody|daemon|\/" | cut -d, -f1 | sed 's|CN=||g'))
 
-if [[ -z "${origHomeDir}" ]]
-then
-	abort "Cannot obtain the original home directory name, is the oldUserName correct?"
+#Ensure old user account is correct and account exists on system
+if [[ ! " ${existingUsers[@]} " =~ " ${oldUser} " ]]; then
+	echo "$oldUser account not present on system to update"
+	exit 1
+fi
+
+#Ensure new user account is not already in use
+if [[ " ${existingUsers[@]} " =~ " ${newUser} " ]]; then
+	echo "$newUser account already present on system. Cannot add duplicate"
+	exit 1
+fi
+
+#Query existing home folders
+existingHomeFolders=($(ls /Users))
+
+#Ensure existing home folder is not in use
+if [[ " ${existingHomeFolders[@]} " =~ " ${newUser} " ]]; then
+	echo "$newUser home folder already in use on system. Cannot add duplicate"
+	exit 1
 fi
 
-dscl . -change /Users/${oldUser} NFSHomeDirectory /Users/${oldUser} /Users/${newUser}
+#Checks if user is logged in
+killProcess=$(ps -Ajc | grep $oldUser | grep loginwindow | awk '{print $2}')
+
+#Logs out user if they are logged in by killing their loginwindow process
+if [[ "${killProcess}" ]]; then
+	sudo kill -9 $killProcess
+	echo "$oldUser account was logged in. User logged off to complete username update."
+fi
+
+#Captures current "RealName" this is the displayName
+fullRealName=$(dscl . -read /Users/${oldUser} RealName)
+
+#Formats "RealName"
+origRealName=$(echo $fullRealName | cut -d' ' -f2-)
+
+#Updates "RealName" to new username (Yes JCAgent will overwrite this after user/system association)
+sudo dscl . -change "/Users/${oldUser}" RealName "${origRealName}" "${newUser}"
+
+err=$?
+if [[ ${err} -ne 0 ]]; then
+	echo "Could not rename the user's RealName in dscl. - err=${err}"
+	exit 1
+fi
+
+#Captures current NFS home directory
+origHomeDir=$(dscl . -read "/Users/${oldUser}" NFSHomeDirectory | awk '{print $2}' -)
+
+if [[ -z "${origHomeDir}" ]]; then
```

Comment:
If this fails, we're already in a condition where changes have been made. It would be best that we check any condition involving a need from oldUser evaluates true before setting changes in motion and/or if any of the steps fail, back them out so the system isn't left in an adverse state.